Map and Partial DAG authoring interface for Dynamic Task Mapping

ashb opened the pull request:
Closes #19889

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

ashb committed:
Map and Partial DAG authoring interface for Operators

ashb committed:
Add TaskGroup mapping DAG author interface

uranusjr commented on tests/models/test_baseoperator.py:
```diff
@@ -600,3 +607,74 @@ class ConcreteSubclassOp(InitSubclassOp, class_arg=class_arg):
+
+    assert task1.downstream_list == [mapped]
+    assert mapped in dag.tasks
+    # At parse time there should only be two tasks!
```
```suggestion
    # At parse time there should only be three tasks!
```

ashb committed:
Add mapping and partial support to TaskFlow tasks

ashb committed:
Add mapping support to `@task_group` decorator

kaxil commented on airflow/decorators/base.py:
```diff
@@ -176,11 +178,92 @@ def _hook_apply_defaults(self, *args, **kwargs):
+        return XComArg(op)
+
+    def map(
+        self, *args, dag: Optional["DAG"] = None, task_group: Optional["TaskGroup"] = None, **kwargs
```
args here isn’t used anywhere. Can we make this function accept keyword only arguments

ashb replied:
Yes, probably.

kaxil commented on airflow/models/dag.py:
```diff
@@ -2197,6 +2197,16 @@ def add_tasks(self, tasks):
             self.add_task(task)
 
+    def _remove_task(self, task_id: str) -> None:
+        # This is "private" as removing could leave a whole in dependencies if done incorrectly, and this
```
```suggestion
        # This is "private" as removing could leave a hole in dependencies if done incorrectly, and this
```

ashb committed:
Improve validation of map and partial argument names

ashb committed:
Remove args from map and parial decorators

uranusjr commented on airflow/decorators/base.py:
```diff
@@ -176,11 +178,92 @@ def _hook_apply_defaults(self, *args, **kwargs):
+@attr.define
+class OperatorWrapper(Generic[T, OperatorSubclass]):
+    """
+    Helper class for providing dynamic task mapping to decorated functions.
```
The name doesn’t really make it clear what this class does, which is more specific than wrapping an operator (although it does). Maybe something like `DecoratedTaskWrapper`?

`task_decorator_factory` might not be a very descriptive name to begin with, but if this is designed most jut for that function, neither `TaskDecoratorFactory` or just `TaskDecorator` are bad names either.

ashb replied:
TaskDecorator sounds good -- this name never really appears outside of this file so maybe `_TaskDecorator`?

uranusjr replied:
That works

ashb replied:
Done in 39b42fa49

ashb committed:
Update airflow/models/dag.py

Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>

uranusjr commented on airflow/decorators/base.py:
```diff
@@ -176,11 +178,110 @@ def _hook_apply_defaults(self, *args, **kwargs):
+
+    @function_arg_names.default
+    def _get_arg_names(self):
+        return set(inspect.signature(self.function).parameters)
```
Does this need to exist? From what I can tell it is only used once in the validator (which already has access to the original `function`).

ashb replied:
`@function_arg_names.default` decorator means that this is the default value for attrs.

No this doesn't need to exist as a function, it could possible be a lamda in `function_arg_names = attr.ib()` call.

ashb replied:
I think a function is clearer than this:

```
    function_arg_names: Set[str] = attr.ib(
        repr=False,
        default=attr.Factory(lambda i: set(inspect.signature(i.function).parameters), takes_self=True),
    )
```

ashb replied:
One option is I could do `del _get_arg_names` -- as once it's decorated the function doesn't need to remain on the class.

ashb replied:
This would be the change that addresses most of your points. What do you think?

```diff
diff --git a/airflow/decorators/base.py b/airflow/decorators/base.py
index 3ee13aed4..6f1fa53b9 100644
--- a/airflow/decorators/base.py
+++ b/airflow/decorators/base.py
@@ -28,6 +28,7 @@ from airflow.models.baseoperator import BaseOperator, MappedOperator
 from airflow.models.dag import DAG, DagContext
 from airflow.models.xcom_arg import XComArg
 from airflow.utils.task_group import TaskGroup, TaskGroupContext
+from airflow.compat.functools import cached_property
 
 
 def validate_python_callable(python_callable):
@@ -181,7 +182,7 @@ T = TypeVar("T", bound=Callable)
 OperatorSubclass = TypeVar("OperatorSubclass", bound="BaseOperator")
 
 
-@attr.define
+@attr.define(slots=False)
 class OperatorWrapper(Generic[T, OperatorSubclass]):
     """
     Helper class for providing dynamic task mapping to decorated functions.
@@ -199,22 +200,32 @@ class OperatorWrapper(Generic[T, OperatorSubclass]):
     decorator_name: str = attr.ib(repr=False, default="task")
     function_arg_names: Set[str] = attr.ib(repr=False)
 
+    @cached_property
+    def function_signature(self):
+        return inspect.signature(self.function)
+
     @function_arg_names.default
     def _get_arg_names(self):
-        return set(inspect.signature(self.function).parameters)
+        return set(self.function_signature.parameters)
+
+    del _get_arg_names
 
     @function.validator
     def _validate_function(self, _, f):
         if 'self' in self.function_arg_names:
             raise TypeError(f'@{self.decorator_name} does not support methods')
 
+    del _validate_function
+
     @multiple_outputs.default
     def _infer_multiple_outputs(self):
-        sig = inspect.signature(self.function).return_annotation
+        sig = self.function_signature.return_annotation
         ttype = getattr(sig, "__origin__", None)
 
         return sig is not inspect.Signature.empty and ttype in (dict, Dict)
 
+    del _infer_multiple_outputs
+
     def __attrs_post_init__(self):
         self.kwargs.setdefault('task_id', self.function.__name__)
 
```

uranusjr replied:
Oh, I meant the `function_arg_names` attribute might not need to exist entirely, since it is only accessed by `_validate_function` (which is a validator for the `function` attribute) and the validator is most likely only run once anyway.

ashb replied:
Ohhh! I thought I used it in .map to validate args then

ashb replied:
It's used in `_validate_arg_names` (which is called from `.map` and `.partial`, and in the function validator, so it probably should exist in some form. It could be a cached property rather than an attr + default.

ashb replied:
See 39b42fa49

uranusjr commented on airflow/decorators/base.py:
```diff
@@ -176,11 +178,110 @@ def _hook_apply_defaults(self, *args, **kwargs):
+        sig = inspect.signature(self.function).return_annotation
+        ttype = getattr(sig, "__origin__", None)
+
+        return sig is not inspect.Signature.empty and ttype in (dict, Dict)
```
Probably better as a cached property? Also the `ttype` line should be `getattr(sig, "__origin__", sig)` because `def foo() -> dict` is still returning multiple outputs.

ashb replied:
I'll double check, but I mostly just moved this code, I didn't change it.

ashb replied:
Yes, this is the old version:

```python
    # try to infer from  type annotation
    if python_callable and multiple_outputs is None:
        sig = signature(python_callable).return_annotation
        ttype = getattr(sig, "__origin__", None)

        multiple_outputs = sig != inspect.Signature.empty and ttype in (dict, Dict)
```

uranusjr commented on airflow/decorators/base.py:
```diff
@@ -176,11 +178,110 @@ def _hook_apply_defaults(self, *args, **kwargs):
+                return
+
+        if len(unknown_args) == 1:
+            raise TypeError(f'{funcname} got unexpected keyword argument {unknown_args.popitem()[0]!r}')
```
I’d do `next(iter(unknown_args))` 🙂

ashb replied:
Done

uranusjr commented on airflow/decorators/base.py:
```diff
@@ -176,11 +178,110 @@ def _hook_apply_defaults(self, *args, **kwargs):
+    ) -> "OperatorWrapper[T, OperatorSubclass]":
+        self._validate_arg_names("partial", kwargs, {'task_id'})
+        partial_kwargs = self.kwargs.copy()
+        partial_kwargs.update(kwargs)
```
This makes me wonder, what should `MyOperator.map(x=something).map(x=another)` do? If I understand this correctly, this would currently discard `something` and just map to `another`. We should likely add something in to prevent this from happening, perhaps in `_validate_arg_names`?

ashb replied:
Yeah, that's an unspecified point of the API. I _think_ I'm leaning towards `.map().map()` being an error on general grounds.

But yes, in my head updating the params was my intent.

potiuk replied:
I also think `map().map()` should be an error. We already agreed on suporting `map(arg1, arg2)` and:

> There should be one-- and preferably only one --obvious way to do it.

ashb replied:
I'm unliekly to be able to make this change before stopping for Christmas, so either someone else can make this, or we can merge it with this and fix it later.

(I think that having a `.map()` function that returns an error would be clearer than having no map method, similar to how I have `.partial()` on a Task object throw an error still.)

uranusjr reviewed (commented):
Not finished, but submitting this batch because my comments already went stale once before I finished.

uranusjr commented on airflow/decorators/task_group.py:
```diff
@@ -20,16 +20,93 @@
+
+    def __del__(self):
+        if not self._invoked:
+            warnings.warn(f"Partial task group {self.function.__name__} was never mapped!")
```
What’s the purpose of this?

ashb replied:
From the tests, to catch this:

```python
    @task_group_decorator
    def tg():
        ...

    with pytest.warns(UserWarning, match='was never mapped'):
        with DAG("test-dag", start_date=DEFAULT_DATE):
            tg.partial()
```

I.e. `tg = tg.partial().map()` was needed. You doesn't make sense to have a partial TG without also mapping it once.

uranusjr replied:
What would happen in this case, nothing? (i.e. this is likely a user mistake)

ashb replied:
Correct -- nothing, and likely a user mistake as the tasks in the task group wouldn't appear in the DAG

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -91,6 +90,23 @@
+    ) -> Callable[..., "MappedOperator"]:
+        # Call this "partial" so it looks nicer in stack traces
+        def partial(*, task_id: str, **kwargs):
+            raise TypeError("partial can only be called on Operator classes, not Tasks themselves")
```
```suggestion
        def partial(**kwargs):
            raise TypeError("partial can only be called on Operator classes, not Tasks themselves")
```

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -196,16 +216,37 @@ def apply_defaults(self, *args: Any, **kwargs: Any) -> Any:
             return result
 
+        apply_defaults.__non_optional_args = non_optional_args  # type: ignore
+        apply_defaults.__param_names = set(non_varaidc_params.keys())  # type: ignore
```
```suggestion
        apply_defaults.__param_names = set(non_varaidc_params)  # type: ignore
```

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -196,16 +216,37 @@ def apply_defaults(self, *args: Any, **kwargs: Any) -> Any:
+        try:
+            # Update the partial descriptor with the class method so it call call the actual function (but let
+            # subclasses override it if they need to)
+            partial_desc = vars(new_cls)['partial']
```
How is this different from `new_cls.partial`?

ashb replied:
```
airflow/models/baseoperator.py:229: in __new__
    partial_desc = new_cls.partial
airflow/models/baseoperator.py:107: in __get__
    return self.class_method.__get__(cls, cls)
E   AttributeError: 'NoneType' object has no attribute '__get__'
```

uranusjr replied:
Ah the descriptor is invoked on the dot syntax. Alright then.

potiuk replied:
Catch 22 :)

ashb replied:
Yeah, this is very much deep stuff that I hope no one ever has to touch again :)

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1659,6 +1629,115 @@ def defer(
+class MappedOperator(DAGNode):
+    """Object representing a mapped operator in a DAG"""
+
+    operator_class: Type[BaseOperator] = attr.ib(repr=lambda c: c.__name__)
```
We should have a validator to ensure `operator_class` and `operator` don’t disagree.

ashb replied:
`operator` attribute isn't actually needed on the instance, so I've changed this approach and added a `from_operator` class method (aka named constructor) that handles it instead.

ashb replied:
See 39b42fa49

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1659,6 +1629,115 @@ def defer(
+    # use a dict so order of args is same as code order
+    unknown_args = value.copy()
+    for clazz in cls.mro():
+        # Mypy doesn't like doing `clas.__init__`, Error is: Cannot access "__init__" directly
```
```suggestion
        # Mypy doesn't like doing `class.__init__`, Error is: Cannot access "__init__" directly
```

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1659,6 +1629,115 @@ def defer(
+        )
+
+
+def _validate_kwarg_names_for_mapping(cls: Type[BaseOperator], func_name: str, value: Dict[str, Any]):
```
I wonder if `_validate_arg_names` and this have similar logic to be refactored out.

ashb replied:
Similar, but the way this gets the possible names (most of the function) is quite differnt, so I couldn't work out a way of really sharing it

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1659,6 +1629,115 @@ def defer(
+        if self.dag and self.operator:
+            # When BaseOperator() was called with a DAG, it would have been added straight away, but now we
+            # are mapped, we want to _remove_ that task (`self.operator`) from the dag
+            self.dag._remove_task(self.task_id)
```
Do we need to do the same for task group? Or is it OK for the unmapped task to stay there?

ashb replied:
It's done inside this function.

uranusjr commented on airflow/models/dag.py:
```diff
@@ -2075,10 +2075,10 @@ def filter_task_group(group, parent_group):
+            group.upstream_group_ids.intersection_update(subdag_task_groups.keys())
+            group.downstream_group_ids.intersection_update(subdag_task_groups.keys())
+            group.upstream_task_ids.intersection_update(dag.task_dict.keys())
+            group.downstream_task_ids.intersection_update(dag.task_dict.keys())
```
```suggestion
            group.upstream_group_ids.intersection_update(subdag_task_groups)
            group.downstream_group_ids.intersection_update(subdag_task_groups)
            group.upstream_task_ids.intersection_update(dag.task_dict)
            group.downstream_task_ids.intersection_update(dag.task_dict)
```

uranusjr commented on tests/serialization/test_dag_serialization.py:
```diff
@@ -1093,14 +1093,12 @@ def test_no_new_fields_added_to_base_operator(self):
         """
         base_operator = BaseOperator(task_id="10")
-        fields = base_operator.__dict__
+        fields = {k: v for (k, v) in vars(base_operator).items() if k in BaseOperator.get_serialized_fields()}
```
Do we have a test to ensure `get_serialized_fields()` is up-to-date? Otherwise we may accidentally add a field on BaseOperator but forget to serialize it, and this test would fail to reflect that.

ashb replied:
We don't, no, this is it. This will pick up anything added via `self.x` in the constructor, but won't pick up things like `x: int = 0` at the class level.

Think it's worth it?

uranusjr commented on airflow/models/taskmixin.py:
```diff
@@ -88,3 +99,167 @@ def __init_subclass__(cls) -> None:
+    dag: Optional["DAG"] = None
+
+    @property
+    @abstractmethod
```
I think this is useless because `DAGNode` (and its parent `DependencyMixin`) does not use `ABCMeta`?

ashb replied:
Fair point -- think I should remove the decorator or change the meta class?

potiuk replied:
I'd vote for adding ABCMeta. The DAGNode becomes an important class for us, and maybe some day we'll add more nodes, who knows.

ashb replied:
Made a real ABC.

uranusjr commented on airflow/models/taskmixin.py:
```diff
@@ -88,3 +99,167 @@ def __init_subclass__(cls) -> None:
+        """Sets relatives for the task or task list."""
+        from airflow.models.baseoperator import BaseOperator, MappedOperator
+
+        if not isinstance(task_or_task_list, Sequence):
+            task_or_task_list = [task_or_task_list]
+
+        task_list: List[DAGNode] = []
+        for task_object in task_or_task_list:
```
Is Mypy OK with this? I’d make `_set_relatives()` only accept the sequence case, and do the coercing outside in the public functions.

ashb replied:
It is, yes.

uranusjr commented on airflow/utils/task_group.py:
```diff
@@ -135,8 +147,9 @@ def __init__(
         self.downstream_group_ids: Set[Optional[str]] = set()
-        self.upstream_task_ids: Set[Optional[str]] = set()
-        self.downstream_task_ids: Set[Optional[str]] = set()
+        # Since the parent class defines these as read-only properties, we can 't just do `self.x = ...`
+        self.__dict__['upstream_task_ids'] = set()
+        self.__dict__['downstream_task_ids'] = set()
```
Can we not change the parent class definition to be plain class-level variable declarations?

ashb replied:
No, because on BaseOperator these are defined as properties.

uranusjr commented on airflow/utils/task_group.py:
```diff
@@ -348,10 +382,35 @@ def build_map(task_group):
+        tg = MappedTaskGroup(self._group_id)
+        tg.mapped_arg = arg
+        tg.mapped_kwargs = {}
+        tg.partial_kwargs = {}
```
Any reason these aren’t in `MappedTaskGroup`’s `__init__`?

ashb replied:
Nope, no idea why they weren't - just fixed.

uranusjr reviewed (commented):
Didn’t read the tests.

ashb committed:
Address some code review comments

ashb committed:
Code review

potiuk commented on airflow/decorators/base.py:
```diff
@@ -199,38 +304,23 @@ def task_decorator_factory(
+    if multiple_outputs is None:
+        multiple_outputs = cast(bool, attr.NOTHING)
+    if python_callable:
+        return _TaskDecorator(  # type: ignore
```
Hmm..  _TaskDecorator is bound to callable so ignore should not be needed I guess? Is the type: ignore here  about the generics

potiuk commented on airflow/models/baseoperator.py:
```diff
@@ -110,12 +126,13 @@ def _apply_defaults(cls, func: T) -> T:
         sig_cache = signature(func)
-        non_optional_args = {
-            name
+        non_varaidc_params = {
```
```suggestion
        non_variadic_params = {
```

potiuk commented on airflow/models/baseoperator.py:
```diff
@@ -110,12 +126,13 @@ def _apply_defaults(cls, func: T) -> T:
+            if param.name != 'self' and param.kind not in (param.VAR_POSITIONAL, param.VAR_KEYWORD)
+        }
+        non_optional_args = {
+            name for (name, param) in non_varaidc_params.items() if param.default == param.empty
```
```suggestion
            name for (name, param) in non_variadic_params.items() if param.default == param.empty
```

potiuk commented on airflow/models/baseoperator.py:
```diff
@@ -196,16 +216,37 @@ def apply_defaults(self, *args: Any, **kwargs: Any) -> Any:
             return result
 
+        apply_defaults.__non_optional_args = non_optional_args  # type: ignore
+        apply_defaults.__param_names = set(non_varaidc_params.keys())  # type: ignore
```
```suggestion
        apply_defaults.__param_names = set(non_variadic_params)  # type: ignore
```

potiuk replied:
And should we be able to mark __paam_names as `Set[str]` ?

ashb replied:
The ignore here is cos of the binding/renaming of `__vars`, not the type of it, so I don't think typing will ever help with `__var` access across class.

potiuk commented on airflow/models/baseoperator.py:
```diff
@@ -792,6 +854,8 @@ def __setattr__(self, key, value):
         if self._lock_for_execution:
             # Skip any custom behaviour during execute
             return
+        if key in self.__init_kwargs:
```
How do we get `self.__init_kwargs` populated ? I could not see it in the change ?

ashb replied:
In apply_defaults:

```python
            if not hasattr(self, '_BaseOperator__init_kwargs'):
                self._BaseOperator__init_kwargs = {}

            result = func(self, **kwargs, default_args=default_args)
            # Store the args passed to init -- we need them to support task.map serialzation!
            self._BaseOperator__init_kwargs.update(kwargs)  # type: ignore
```

ashb committed:
Move initializers to MappedTaskGroup constructor

potiuk reviewed (commented):
First pass but it will require a few more :) to understand all intricacies :)

ashb committed:
Make DAGNode a proper Abstract Base Class

ashb committed:
Apply suggestions from code review

Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

uranusjr committed:
Prevent mapping an already mapped Task/TaskGroup

Also prevent calls like .partial(...).partial(...). It is uncertain
whether these kinds of repeated partial/map calls have utility, so let's
disable them entirely for now to simplify implementation. We can always
add them if they are proven useful.

ashb commented on airflow/decorators/task_group.py:
```diff
@@ -90,20 +88,26 @@ def _make_task_group(self, **kwargs) -> MappedTaskGroup:
 
     def __del__(self):
-        if not self._invoked:
+        if not self.mapped_kwargs:
```
Edge case (that I _think_ is fine): This would mean that `tg.partial(a=2).map()` wouldn't work -- as the mapped_kwargs would be an empty dict and evaluate to false.

uranusjr replied:
Maybe we should even disable `map()` and `partial()` without any arguments (because how’d they make sense…?) but let’s not bother.

ashb committed:
fixup! Add mapping and partial support to TaskFlow tasks

ashb commented on airflow/decorators/base.py:
```diff
@@ -275,10 +275,10 @@ def map(
         self, *, dag: Optional["DAG"] = None, task_group: Optional["TaskGroup"] = None, **kwargs
     ) -> "_TaskDecorator[T, OperatorSubclass]":
+        if self.kwargs:
+            raise RuntimeError("Already a partial task")
```
Hmmm, this one doesn't behave correctly right now, as:

```
@taks(task_id='x')
def y():
    ...
```

This currently stores `task_id` in the `self.kwargs`

ashb replied:
Removed this check only in deb54df1c

ashb committed:
fixup! Prevent mapping an already mapped Task/TaskGroup

ashb committed:
fixup! Add mapping and partial support to TaskFlow tasks

ashb committed:
fixup! Add mapping support to `@task_group` decorator
